### PR TITLE
Implemented private bicep registry publish

### PR DESCRIPTION
**Pull request by MrMCake**
# Change

- Implemented private bicep registry publish
- Added registry to dependency pipeline (including reference of `artifacts-rg`)
- Added module usage example to docs
- Streamlined pipeline parameter names
- Aligned template spec module publish name with bicep registry
  - I.e. Instead of publishing 'servers' we now publish 'microsoft.sql.servers'
  
## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update (Wiki)

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1 [hunk 1]
```diff
@@ -10,19 +10,20 @@ Otherwise, one of the provided version options is chosen and applied with the de
...
 
 .PARAMETER templateFilePath
 Mandatory. Path to the module deployment file from root.
 
-.PARAMETER componentTemplateSpecRGName
+.PARAMETER templateSpecsRGName
 Mandatory. ResourceGroup of the template spec to publish to.
 
-.PARAMETER componentTemplateSpecRGLocation
+.PARAMETER templateSpecsRGLocation
```
[the remark this answers is not in the document]
Comment (MrMCake) [reply .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1]: updated

File: .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1 [hunk 2]
```diff
@@ -10,19 +10,20 @@ Otherwise, one of the provided version options is chosen and applied with the de
...
-Mandatory. The name of the module to publish. It will be the name of the template spec.
+.PARAMETER moduleIdentifier
+Mandatory. The identifier of the module to publish (ProviderNamespace/ResourceType Combination). It will be the name of the template spec.
+E.g. 'Microsoft.KeyVault/vaults'
 
 .PARAMETER templateFilePath
 Mandatory. Path to the module deployment file from root.
 
-.PARAMETER componentTemplateSpecRGName
+.PARAMETER templateSpecsRGName
```
[the remark this answers is not in the document]
Comment (MrMCake) [reply .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1]: updated

File: .github/actions/templates/publishModule/scripts/Publish-ModuleToPrivateBicepRegistry.ps1 [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    [CmdletBinding(SupportsShouldProcess)]
+    param (
+        [Parameter(Mandatory)]
+        [string] $moduleIdentifier,
+
+        [Parameter(Mandatory)]
+        [string] $templateFilePath,
+
+        [Parameter(Mandatory)]
+        [string] $bicepRegistryRGName,
```
[the remark this answers is not in the document]
Comment (MrMCake) [reply .github/actions/templates/publishModule/scripts/Publish-ModuleToPrivateBicepRegistry.ps1]: updated

File: .github/actions/templates/publishModule/action.yml [hunk 4]
```diff
@@ -5,29 +5,36 @@ inputs:
...
-  versioningOption:
-    description: 'The mode to handle the version increments [major|minor|patch|custom].'
-    default: 'patch'
+  bicepRegistryName:
+    description: 'Required to publish to private bicep registry. ResourceGroup of the container registry to publish to'
     required: false
-  customVersion:
-    description: 'The version to enforce if [versioningOption] is set to [custom]'
-    default: '0.0.1'
+  bicepRegistryRGName:
```
[the remark this answers is not in the document]
Comment (MrMCake) [reply .github/actions/templates/publishModule/action.yml]: updated

File: .github/actions/templates/publishModule/scripts/Publish-ModuleToPrivateBicepRegistry.ps1 [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        if (-not ([String]::IsNullOrEmpty($customVersion)) -and ((New-Object System.Version($customVersion)) -gt (New-Object System.Version($latestVersion)))) {
+            Write-Verbose "A custom version [$customVersion] was specified in the pipeline script and is higher than the current latest. Using it."
+            $newVersion = $customVersion
+        } else {
+            Write-Verbose 'No custom version set. Using default versioning.'
+
+            switch ($versioningOption) {
+                'major' {
+                    Write-Verbose 'Apply version update on "major" level'
+                    $newVersion = (New-Object -TypeName System.Version -ArgumentList ($latestVersion.Major + 1), $latestVersion.Minor, $latestVersion.Build).ToString()
```
Comment (MariusStorhaug) [inline .github/actions/templates/publishModule/scripts/Publish-ModuleToPrivateBicepRegistry.ps1:105]: Patch and minor versions MUST be reset to 0 when major version is incremented.

https://semver.org/#spec-item-8

```suggestion
                    $newVersion = (New-Object -TypeName System.Version -ArgumentList ($latestVersion.Major + 1), 0, 0).ToString()
```

File: .github/actions/templates/publishModule/scripts/Publish-ModuleToPrivateBicepRegistry.ps1 [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+                    break
+                }
+                'minor' {
+                    Write-Verbose 'Apply version update on "minor" level'
+                    $newVersion = (New-Object -TypeName System.Version -ArgumentList $latestVersion.Major, ($latestVersion.Minor + 1), $latestVersion.Build).ToString()
```
Comment (MariusStorhaug) [inline .github/actions/templates/publishModule/scripts/Publish-ModuleToPrivateBicepRegistry.ps1:110]: Patch version MUST be reset to 0 when minor version is incremented.

https://semver.org/#spec-item-7

```suggestion
                    $newVersion = (New-Object -TypeName System.Version -ArgumentList $latestVersion.Major, ($latestVersion.Minor + 1), 0).ToString()
```

File: .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1 [hunk 7]
```diff
@@ -17,10 +17,10 @@ Mandatory. The name of the module to publish. It will be the name of the templat
 Mandatory. Path to the module deployment file from root.
 
 .PARAMETER componentTemplateSpecRGName
-Mandatory. ResourceGroup of the template spec to publish to.
+Mandatory. ResourceGroupRgf the template spec to publish to.
```
Comment (MariusStorhaug) [inline .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1:20]: Not sure what this is. Seems a search and replace gone wrong?

File: .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1 [hunk 8]
```diff
@@ -17,10 +17,10 @@ Mandatory. The name of the module to publish. It will be the name of the templat
...
 
 .PARAMETER componentTemplateSpecRGLocation
-Mandatory. Location of the template spec resource group.
+Mandatory. Location of tRg template spec resource group.
```
Comment (MariusStorhaug) [inline .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1:23]: Not sure what this is. Seems a search and replace gone wrong?

File: .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1 [hunk 9]
```diff
@@ -72,7 +72,7 @@ function Publish-ModuleToTemplateSpec {
         ##    EVALUATE RESOURCES   ##
         #############################
         if (-not (Get-AzResourceGroup -Name $componentTemplateSpecRGName -ErrorAction 'SilentlyContinue')) {
-            if ($PSCmdlet.ShouldProcess("Resource group [$componentTemplateSpecRGName] to location [$componentTemplateSpecRGLocation]", 'Deploy')) {
+            if ($PSCmdlet.ShouldProcess("Resource group [$RgmponentTemplateSpecRGName] to location [$componentTemplateSpecRGLocation]", 'Deploy')) {
```
Comment (MariusStorhaug) [inline .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1:75]: Not sure what this is. Seems a search and replace gone wrong?

File: .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1 [hunk 10]
```diff
@@ -135,11 +135,11 @@ function Publish-ModuleToTemplateSpec {
         if ($PSCmdlet.ShouldProcess("Template spec [$componentTemplateSpecName] version [$newVersion]", 'Publish')) {
             $templateSpecInputObject = @{
                 ResourceGroupName = $componentTemplateSpecRGName
-                Name              = $componentTemplateSpecName
+                Name              = $componentTempRgteSpecName
```
Comment (MariusStorhaug) [inline .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1:138]: Not sure what this is. Seems a search and replace gone wrong?

File: .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1 [hunk 11]
```diff
@@ -135,11 +135,11 @@ function Publish-ModuleToTemplateSpec {
...
                 Version           = $newVersion
                 Description       = $componentTemplateSpecDescription
                 Location          = $componentTemplateSpecRGLocation
-                TemplateFile      = $templateFilePath
+                TemplateFile      = $templateFilePRgh
```
Comment (MariusStorhaug) [inline .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1:142]: Not sure what this is. Seems a search and replace gone wrong?

Comment (MariusStorhaug) [reply .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1:142]: Have you had 'insert' enabled?

File: .github/actions/templates/publishModule/action.yml [hunk 12]
```diff
@@ -75,3 +87,36 @@ runs:
...
+            versioningOption    = '${{ inputs.versioningOption }}'
+          }
+
+          Write-Verbose "Invoke task with" -Verbose
+          Write-Verbose ($functionInput | ConvertTo-Json | Out-String) -Verbose
+
+          Publish-ModuleToPrivateBicepRegistry @functionInput -Verbose
+        } else {
+          Write-Output "Pipeline is configured to not publish to TemplateSpecs"
+        }
```
Comment (MariusStorhaug) [inline .github/actions/templates/publishModule/action.yml:122]: Could we move this to a PowerShell file and just call it directly as a single command in run? i.e. main.ps1. If its in PowerShell we get better support in tools. Maybe this restructure could happen separately, but I'm more a fan of having code in file where we actually can get support from the tools we are using. It goes for all the places we have workflows backed by powershell (or other). As a reference. Look at the [Azure/Login action.yml](https://github.com/Azure/login/blob/master/action.yml) and [action/checkout](https://github.com/actions/checkout/blob/main/action.yml)

Comment (MariusStorhaug) [reply .github/actions/templates/publishModule/action.yml:122]: Not sure how you would get the inputs from a PowerShell context however, but maybe it just works using `${{ input.something }}`-syntax in the PowerShell file.

**Review by MariusStorhaug — CHANGES_REQUESTED** [review]

File: .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1 (hunk 7 above)
Comment (MrMCake) [reply .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1]: seems to be fixed

File: .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1 (hunk 8 above)
Comment (MrMCake) [reply .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1]: seems to be fixed

File: .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1 (hunk 9 above)
Comment (MrMCake) [reply .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1]: seems to be fixed

File: .github/actions/templates/publishModule/action.yml (hunk 12 above)
Comment (MariusStorhaug) [reply .github/actions/templates/publishModule/action.yml:122]: Opened another issue for this: #427

File: .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1 (hunk 10 above)
Comment (MrMCake) [reply .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1]: seems to be fixed

File: .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1 (hunk 11 above)
Comment (MrMCake) [reply .github/actions/templates/publishModule/scripts/Publish-ModuleToTemplateSpec.ps1]: seems to be fixed

**Review by MariusStorhaug — APPROVED** [review]
Seems all comments are fixed. Approved!

**Pull request merged**